### ISSUE TITLE
[RFE] Steve equivalents for login / logout

### ISSUE THREAD

**Issue by richard-cox** [rancher/rancher#47865]
**Is your feature request related to a problem? Please describe.**
- In order to support SURE-6548 (Public API: Tokens) the UI must move away from the norman endpoints used to login 
  - Norman endpoints do not use the new token
- Unfortunately login is very much tied to the auth providers which 
  - norman supplied list of auth providers - `/v3-public/authproviders` contains url for...
  - norman login in `/v3-public/localProviders/local?action=login"`
    - which returns the legacy token
    - there's probably a LOT more funkier/norman things going on for 3rd party auth providers
- log out may be for free via the public api tokens changes. it currently uses `/v3/tokens?action=logout`

**Describe the solution you'd like**
- steve support for the login process, including third party auth providers
- steve support should utilise the new public api tokens

**Additional context**
- blocks https://github.com/rancher/dashboard/issues/12331

**Comment by richard-cox** [comment]
Potentially blocked on https://github.com/rancher/rancher/issues/47033 (any new public/imperative api would need to be accessible via steve by the ui)

**Comment by andreas-kupries** [comment]
Talked to @raulcabello about this, and plan to snarf his ideas from https://github.com/raulcabello/rancher-architecture/blob/user-public-api/docs/RFD%200011%20User%20Public%20API.md#new-resources

Basically create minimal imperative API resources supporting only `create`, and no other methods.
One resource per request, i.e. one for login, and one for logout.
These receive the action request and return the request extended with result data (`Status` field) from the action.
Nothing is stored, fully ephemeral, only the side-effects of the action itself matter.

🚧 Locate the norman login/logout input/output structures to repro to study and repro analogously in the new resources.

**Comment by andreas-kupries** [comment]
## login action

pkg/apis/management.cattle.io/v3/public_authn_types.go: `type *Login struct { ... }`

### v3 structures
```
 GenericLogin (ttl, description, response type (json|cookie))
   AzureADLogin (code, id token)
   BasicLogin (username, password)
   GenericOIDCLogin (code)
   GithubLogin (code)
   GoogleOauthLogin (code)
   OIDCLogin (code)

 SamlLoginInput (final redir url, req-id, public key, response type)
 SamlLoginOutput (redir url)
```
### response types

- `cookie`: empty body, token is set as cookie on the http response
- `saml`: empty body, special handling, see the `SamlLogin*` structures
- else: json object, key `token`, value is string usable for auth (`<name>:<secretvalue>`)

I am not sure if the cookie response type would fit into the steve imp API. I do not believe that we have that much access to the http response for the request. Solvable tough, simply handle all as the status of the pseudo resource.

:boom: ⚠ The `saml` response type is a __problem__ however. This is internally generated to indicate a saml provider. These have a very custom flow, with multiple redirections through the browser/client until the backend receives a SAML assertion from which it can make the token. Which is then unconditionally set as cookie on the final http response.

I am not sure how to fit this into the interfaces we have from the imp API. 

@tomleb do you have ideas, comments, thoughts about the above ?

@raulcabello I suspect that the user helper resources do not have these complexities.

### endpoints

With norman each auth provider has its own url/endpoint for the login action (*). This passes the provider type to the action without having it mentioned in the login request itself. In the backend the information is known through the `request.Type` (field in the norman `APIContext`) accessible by the shared action code.

(Thanks to @richard-cox for the information)

(*) See the result of the `/v3-public/authProviders` endpoint.

For public API / steve this should be a single endpoint, and the auth provider type part of the pseudo resource request.
The dashboard can pull the type information out of the `id` field, instead of `actions.login`.

**Comment by andreas-kupries** [comment]
## logout action

pkg/auth/tokens/manager.go `func (m *Manager) logout ...`

The action mostly does two things:
1. delete the token
2. unset the auth cookie (`R_SESS`).

With steve this could be done as regular delete of the token, plus unsetting the cookie from the dashboard.
❓ Actually is the latter possible ?
❓ Analogously for `login`, can the dashboard set the auth cookie after it got the token info from the action ?

:boom: SAML is again the problem, with its support of `logoutAll`. This again performs a complex flow with multiple redirections through the browser/client until the backend receives the logout done and can return to the browser with the final redirection.

**Comment by pmatseykanets** [comment]
It seems like we'll need to implement steve handlers for what can be `/v1-public` if we have to move away from norman. This should solve redirect flows, cookies, and public api (as in public api endpoints). Internally login/logout endpoints can delegate to the corresponding imperative resources e.g. `LoginRequest/LogoutRequest` if we want to but that's not necessarily required IMHO.
We also need to coordinate with AuthConfig Public API work to ensure consistency of the overall API experience (type names, values etc.).

**Comment by pmatseykanets** [comment]
Proposed changes:

1. `GET /v1-public/authproviders` lists enabled auth providers without any decorations (links etc.). There are always only either one (local) or two entries returned (local and a third party provider).

```json
{
    "data": [
        {
            "id": "local",
            "logoutAllEnabled": false,
            "logoutAllForced": false,
            "logoutAllSupported": false,
            "type": "localProvider"
        },
        {
            "id": "okta",
            "logoutAllEnabled": true,
            "logoutAllForced": true,
            "logoutAllSupported": true,
            "redirectUrl": "…/v1-saml/okta/login",
            "type": "oktaProvider"
        }
    ]
}
```
Only the following common fields are returned: `id`, `type`, `logoutAllEnabled`, `logoutAllForced`, `logoutAllSupported`. All provider-specific fields are returned as before.

2. There will be no `GET /v1-public/<authProviderType>/<id>` endpoints as everything needed can be obtained from n.1

3. `POST /v1-public/login` is a unified login endpoint that replaces existing `POST /v3-public/<providerType>/<id>?action-login` endpoints that requires to pass the auth provider type in the payload:

```json
{
    "type": "localProvider",
    "description": "UI session",
    "responseType": "cookie",
    "username": "...",
    "password": "..."
}
```

4. `GET /v1-public/authtokens/<id>` and `DELETE /v1-public/<id>` behave as before
5. `POST /v1/logout[?all]` that behaves the same way as `/v3/tokens?action={logout|logoutAll}`

**Comment by tomleb** [comment]
> Internally login/logout endpoints can delegate to the corresponding imperative resources e.g. LoginRequest/LogoutRequest if we want to but that's not necessarily required IMHO.

Would be great if this ended up being part of Public API so that users using `/v3` have a way to migrate. For example, https://github.com/fidelity/kconnect/blob/d93cd58c15e5d17dc350e9b849ff4a6ff07ddd60/pkg/rancher/endpoints.go#L25 currently using /v3 to login to via active directory provider. Have we discussed this somewhere, do we want to support this (perhaps as part of another ticket?).

**Comment by pmatseykanets** [comment]
## Validation Template

## What was fixed, or what change has occurred

-  Added `/v1-public` endpoints side-by-side with `/v3-public` ones:
    - `GET /v1-public/authproviders`
    - `POST /v1-public/login`
    - `GET /v1-public/authtokens`
    - `DELETE /v1-public/authtokens`
- Added `/v1/logout[?all]` endpoint side-by-side with `/v3/tokens?action={logout|logoutAll}`.
- Refactored token and user manager's code and adjacent call sites to leave token related concerns to the token manager and user ones to the user manager.

## Areas or cases that should be tested

- Existing `/v3-public` endpoints
    - `GET /v3-public/authProviders`
    - `POST /v3-public/login/<authProviderType>s/<id>?action=login` e.g. `/v3-public/localProviders/local?action=login`. Ideally for all auth providers.
    - `GET /v3-public/authtokens`
    - `DELETE /v3-public/authtokens`
- New `/v1-public` endpoints
    - `GET /v1-public/authproviders`
    - `POST /v1-public/login`
    - `GET /v1-public/authtokens`
    - `DELETE /v1-public/authtokens`
- Existing `/v3/tokens?action={logout|logoutAll}` endpoint
- New `/v1/logout[?all]` endpoint
- Group membership refresh.Ideally for all refreshable auth providers.
- Rancher CLI `token` command that uses the `/v3-public/authtokens` endpoints.
- Rancher terraform provider is able to login.

## What areas could experience regressions

- Issues with logging in for some or all auth providers
- Issues with logging out for some or all auth providers
- Group membership refresh

## Are the repro steps accurate/minimal?

N/A.

**Comment by dasarinaidu** [comment]
**Observation 1:  POST  `{{rancher_url}}/v1-public/login` fails for OAuth providers(azureAD) with 500 error**

I connected my rancher to AzureAD and try to hit below endpoint from postman

Endpoint: POST /v1-public/login
Request Body:
```
json{
  "type": "azureADProvider",
  "responseType": "json"
}
```
**Expected Behavior:**

- Return 200 OK with redirect URL to Azure AD, OR
- Return 302 redirect to Azure AD login page

**Actual Behavior:**

- Returns 500 Internal Server Error
- Server log shows: The request body must contain the following parameter: 'code'

**Root Cause:**
The endpoint is trying to process an OAuth callback (exchange code for token) instead of initiating the OAuth flow.
**Comparison:**
The v3 endpoint (GET /v3-public/azureADProviders/azuread?action=login) correctly returns the authUrl for OAuth initiation.
**Impact:**
Cannot test Azure AD authentication via the new v1-public/login endpoint. v3 endpoint still works.

NOTE: This endpoint works with OpenLDAP; i m getting 201



#### Test 1: Azure AD Provider

**Request:**
```
POST /v1-public/login
{
  "type": "azureADProvider",
  "responseType": "json"
}
```

**Response:** 500 Internal Server Error

**Rancher Server Log:**
```
2025/11/06 23:13:00 [ERROR] login: Error authenticating user: getting OID from AuthCode: could not retrieve token from auth code: http call(https://login.microsoftonline.com/1a9db2ee-2bb5-42a6-9bbe-c819656fcb39/oauth2/v2.0/token)(POST) error: reply status code was 400:
{"error":"invalid_request","error_description":"AADSTS900144: The request body must contain the following parameter: 'code'. Trace ID: 373edf2a-ac57-4d9e-8508-27034a793400 Correlation ID: e5118d83-7d0d-46c7-a52a-21309d226d54 Timestamp: 2025-11-06 23:13:00Z","error_codes":[900144],"timestamp":"2025-11-06 23:13:00Z","trace_id":"373edf2a-ac57-4d9e-8508-27034a793400","correlation_id":"e5118d83-7d0d-46c7-a52a-21309d226d54","error_uri":"https://login.microsoftonline.com/error?code=900144"}
```

**Analysis:** Server attempted to exchange OAuth authorization code with Azure AD, but no code was provided in the request.

---

#### Test 2: Amazon Cognito Provider

**Request:**
```
POST /v1-public/login
{
  "type": "cognitoProvider"
}
```

**Response:** 500 Internal Server Error

**Rancher Server Log:**
```
2025/11/06 23:32:41 [ERROR] login: Error authenticating user: oauth2: "invalid_request" "invalid_code"
2025/11/06 23:35:03 [ERROR] login: Error authenticating user: oauth2: "invalid_request" "invalid_code"
```

**Analysis:** Server attempted to validate OAuth authorization code with Cognito, but received "invalid_code" error indicating no valid code was provided.

**Comment by dasarinaidu** [comment]
<html><head></head><body><h2>Observation 2: Logout-All Fails to Invalidate OAuth Tokens</h2>
<hr>
<h3><strong>Description</strong></h3>
<p><code>POST /v1/logout?all=true</code> fails to invalidate tokens when used with OAuth provider authentication (Cognito, Azure AD). The endpoint returns 500 error and the token remains active and usable.</p>
<hr>
<h3><strong>Steps to Reproduce</strong></h3>
<ol>
<li>Login via OAuth provider (Cognito or Azure AD) using browser</li>
<li>Extract authentication token from browser session</li>
<li>Verify token is active:
<pre><code>GET /v3/users?me=trueAuthorization: Bearer [oauth-token]Result: 200 OK ✓
</code></pre></li>
<li>Call logout-all:
<pre><code>POST /v1/logout?all=trueAuthorization: Bearer [oauth-token]Result: 500 Internal Server Error
</code></pre></li>
<li>Verify token status:
<pre><code>GET /v3/users?me=trueAuthorization: Bearer [oauth-token]Result: 200 OK (token still active)
</code></pre></li>
</ol>
<hr>
<h3><strong>Expected Behavior</strong></h3>
<ul>
<li>Returns 200 OK or 204 No Content</li>
<li>Token is invalidated</li>
<li>Subsequent API calls with token return 401 Unauthorized</li>
</ul>
<hr>
<h3><strong>Actual Behavior</strong></h3>
<ul>
<li>Returns 500 Internal Server Error</li>
<li>Token remains active and usable</li>
<li>Subsequent API calls with token still return 200 OK</li>
<li>User is NOT logged out</li>
</ul>
<hr>
<h3><strong>Server Logs</strong></h3>
<pre><code>2025/11/07 03:22:52 [ERROR] logout: provider logout: InvalidBodyContent 422: CognitoProvider: parsing request body: EOF
</code></pre>
<p><strong>Analysis:</strong> Server attempts to notify OAuth provider with invalid/empty request body, provider rejects with 422, causing entire logout operation to fail and token to remain active.</p>
<hr>
<h3><strong>Additional Information</strong></h3>
<p><strong>Affected Endpoints:</strong></p>
<ul>
<li><code>POST /v1/logout?all=true</code> - Fails with OAuth tokens</li>
<li><code>POST /v3/tokens?action=logoutAll</code> - Fails with OAuth tokens (same issue)</li>
</ul>
<p><strong>Working Endpoints:</strong></p>
<ul>
<li><code>POST /v1/logout</code> (single logout) - Works with OAuth tokens ✓</li>
<li><code>POST /v1/logout?all=true</code> - Works with local provider tokens ✓</li>
</ul>
<p><strong>Test Matrix:</strong></p>

Endpoint | Local Token | OAuth Token
-- | -- | --
/v1/logout | ✅ 200, token invalidated | ✅ 200, token invalidated
/v1/logout?all=true | ✅ 200, token invalidated | ❌ 500, token STILL ACTIVE
/v3/tokens?action=logout | ✅ 200, token invalidated | ✅ 200, token invalidated
/v3/tokens?action=logoutAll | ✅ 200, token invalidated | ❌ 500, token STILL ACTIVE

**Comment by dasarinaidu** [comment]
**Observation 2: v1-public GET endpoint throws 404**

I connected to keycloak SAML and trying to fetch the authtokens and getting 404 error

❌ GET /v1-public/authtokens returns 404 (list tokens doesn't work)
❌ GET /v1-public/authtokens/{id} returns 404 (get specific token doesn't work)


**Actual:**
I am getting 404 error on all the token endpoints


**Observation 3: v1-public DELETE endpoint is throwing 200 but the token is not deleted**

⚠️ DELETE /v1-public/authtokens/{id} returns 200 OK, but:

Returns 200 on multiple DELETE attempts (should return 404 after first delete)

**Comment by pmatseykanets** [comment]
> GET /v1-public/authtokens returns 404 (list tokens doesn't work)

This is by design. Allowing listing auth tokens would be a security issue.

> GET /v1-public/authtokens/{id} returns 404 (get specific token doesn't work)

How do you test for it? The resource behind this endpoint `samltoken` is normally short-lived one e.g. when the auth flow is initiated by rancher cli (`rancher token` command). You can create the resource directly though and see if the endpoint returns it e.g.

```
kubectl create -f -<<EOF
apiVersion: management.cattle.io/v3
kind: SamlToken
metadata:
  name: "token-12345"
  namespace: "cattle-global-data"
EOF
samltoken.management.cattle.io/token-12345 created

kubectl get samltoken -n cattle-global-data
NAME          AGE
token-12345   81s

curl -s https://<rancher-url>/v1-public/authtokens/token-12345
{"expiresAt":"","token":""}

curl -s -X DELETE https://<rancher-url>/v1-public/authtokens/token-12345

kubectl get samltoken -n cattle-global-data
No resources found in cattle-global-data namespace.
```

> DELETE /v1-public/authtokens/{id} returns 200 OK, but:
> Returns 200 on multiple DELETE attempts (should return 404 after first delete)

This is by design. Deleting an auth token is an idempotent operation.

**Comment by dasarinaidu** [comment]
**Observation 3: Logout endpoint returns 500 Internal Server Error when no authentication token is provided, instead of 401 Unauthorized.**

POST {{rancher_url}}/v1/logout
(No Authorization header)

```
Body:
{
  "finalRedirectUrl": "..."
}
```

**Expected Behavior**

Status: 401 Unauthorized
Message: "Authentication required" or "Unauthorized"

**Actual Behavior**

Status: 500 Internal Server Error
Message: "Internal Server Error"


Rancher Logs:
```
2025/11/11 17:54:55 [ERROR] logout: provider logout: SAML [logout]: Rancher provider resource `keycloak` configured for forced SLO, rejecting regular logout
```

**Comment by pmatseykanets** [comment]
> Observation 3: Logout endpoint returns 500 Internal Server Error when no authentication token is provided, instead of 401 Unauthorized.

Doesn't seem to be reproducable.

> SAML [logout]: Rancher provider resource `keycloak` configured for forced SLO, rejecting regular logout

This is expected behavior conveying misconfiguration/misuse of the logout feature. If you configured forced logout you can't request to simply logout out of Rancher, you have to use logout all - `/v1/logout?all`.

**Comment by dasarinaidu** [comment]
Thanks @pmatseykanets ; I tried with `/v1/logout?all` and i am getting 401

**Comment by pmatseykanets** [comment]
> I tried with /v1/logout?all and i am getting 401

You should have the same outcome 401 for both `/v1/logout` and `/v1/logout?all` when not authenticated.

**Comment by dasarinaidu** [comment]
Thanks @pmatseykanets. I validated with curl and both endpoints giving me 401.
 
All the tests are GREEN and closing this card. 

| # | Test Case | Provider | Status | Result |
|---|-----------|----------|--------|--------|
| 1 | GET /v1-public/authproviders | All | PASSED | Returns list of configured auth providers |
| 2 | GET /v1-public/authproviders/{id} | All | PASSED | Returns individual provider details |
| 3 | POST /v1-public/login - Valid credentials | Local | PASSED | Returns auth token successfully |
| 4 | POST /v1-public/login - Invalid credentials | Local | PASSED | Returns 401 Unauthorized with error message |
| 5 | POST /v1-public/login - Legacy /v3 | Local | PASSED | Backward compatibility maintained |
| 6 | POST /v1-public/login - OAuth providers | Azure AD, Google, Cognito | BLOCKED | Requires browser OAuth flow (by design) |
| 7 | POST /v1/logout - Single session | Keycloak SAML | PASSED | Session terminated, token invalidated |
| 8 | POST /v1/logout - Token validation | Keycloak SAML | PASSED | Token invalid after logout (401 on reuse) |
| 9 | POST /v1/logout?all=true - All sessions | Keycloak SAML | PASSED | All sessions terminated with finalRedirectUrl |
| 10 | POST /v1/logout?all=true - Forced SLO | Keycloak SAML | PASSED | Forced SLO requires ?all=true endpoint |
| 11 | POST /v3/tokens?action=logout | Keycloak SAML | PASSED | Legacy logout works correctly |
| 12 | POST /v3/tokens?action=logoutAll | Keycloak SAML | PASSED | Legacy logout-all works with finalRedirectUrl |
| 13 | DELETE /v1-public/authtokens/{id} | Local | PASSED | Token deleted successfully (idempotent) |
| 14 | GET /v3/tokens - List tokens | Local | PASSED | Returns active token list (v1 GET disabled for security) |
| 15 | Token deletion verification | Local | PASSED | Deleted token not in list |
| 16 | Token compatibility v1↔v3 | Local | PASSED | v1 tokens work with v3 APIs |
| 17 | Invalid provider type | N/A | PASSED | Returns 422 InvalidBodyContent |
| 18 | Missing username field | Local | PASSED | Returns 401 authentication failed |
| 19 | Missing password field | Local | PASSED | Returns 401 authentication failed |
| 20 | Missing type field | N/A | PASSED | Returns 422 InvalidBodyContent |
| 21 | Empty request body | N/A | PASSED | Returns 422 InvalidBodyContent |
| 22 | Malformed JSON | N/A | PASSED | Returns 422 InvalidBodyContent |
| 23 | Logout without auth token | Keycloak SAML | PASSED | Returns 401 Unauthorized (with ?all=true) |
| 24 | Logout with malformed auth header | Keycloak SAML | PASSED | Returns 401 Unauthorized |
| 25 | Logout with invalid token | Keycloak SAML | PASSED | Returns 401 must authenticate |
| 26 | SLO - Okta SAML | Okta SAML | PASSED | Session terminated, re-login requires credentials |
| 27 | SLO - Okta OIDC | Okta OIDC | PASSED | Session terminated, re-login requires credentials |
| 28 | SLO - Azure AD OIDC | Azure AD OIDC | PASSED | Session terminated, re-login requires credentials |
| 29 | SLO - Amazon Cognito | Amazon Cognito | PASSED | Session terminated, re-login requires credentials |

**Summary:**
- Total Tests: 29
- Passed: 26
- Blocked: 3 (OAuth login - by design)
- Failed: 0
- Pass Rate: 100%
